5주차 과제를 제출합니다

Kwakcena opened the pull request:
5주차 과제를 제출합니다

Kwakcena committed:
CORS 에러를 해결하라

ProductController 에 CrossOrigin 어노테이션을 붙여 CORS 에러를 해결합니다

Kwakcena committed:
User model을 만들어라

회원에 대한 model entity를 만들었습니다

Kwakcena committed:
사용자를 생성하는 controller 메서드를 만들어라

사용자를 생성하는 요청인 POST /users 에 대한 메소드를 만들고, 이를 테스트 합니다.
필수 항목이 들어오지 않으면 Bad request를 보내는 테스트도 작성하였습니다.

Kwakcena committed:
UserNotFoundException을 만들어라

존재하지 않는 ID의 user를 요청할 때 반환할 에러의 클래스를 만들고 이를 스프링으로 관리하여 반환하게 합니다.

Kwakcena committed:
User update 요청을 구현하라

UserController로 user의 정보 변경에 대한 요청 기능을 만들고, 이에 대한 테스트를 작성하였습니다

Kwakcena committed:
UserController의 회원 삭제 기능을 구현하라

회원 정보를 삭제하는 기능을 구현하고 이를 테스트 하였습니다.

Kwakcena committed:
@BeforeEach를 사용하여 테스트 전에 정의하라

테스트에 사용되는 given 같은 것들은 BeforeEach를 사용하여 정의할 수 있도록 한다

Kwakcena commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -191,8 +199,9 @@
             @DisplayName("Not found 예외를 던진다")
-            void 회원_정보를_삭제한다() throws Exception {
-                given(userService.delete(1000L)).willThrow(new UserNotFoundException(1000L));
+            void Not_found_예외를_던진다() throws Exception {
+                given(userService.delete(1000L))
+                        .willThrow(new UserNotFoundException(1000L));
```
`deleteUser` 에 대한 테스트 중 존재하지 않는 ID에 대해 삭제 요청을 하면 예외를 던지는 테스트가 있는데,

여기서 사용하는 given을 BeforeEach로 올리면 `User not found: 1000` 메세지가 뜨면서 테스트가 실패합니다.

update와는 다르게 여기서는 실패하는 이유를 잘 모르겠습니다.

어느 부분에서 놓치고 있는지 조언 부탁드리겠습니다
```java
@Nested
@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
class deleteUser_메소드는 {

    @BeforeEach
    void setUp() {
        given(userService.delete(1000L))
                .willThrow(new UserNotFoundException(1000L));
    }

    @Nested
    @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
    class ID가_존재하는_회원이라면 {

        @Test
        @DisplayName("회원 정보를 삭제한다")
        void 회원_정보를_삭제한다() throws Exception {
            mockMvc.perform(delete("/users/1"))
                    .andExpect(status().isNoContent());

            verify(userService).delete(1L);
        }
    }

    @Nested
    @DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
    class ID가_존재하지않는_회원이라면 {

        @Test
        @DisplayName("Not found 예외를 던진다")
        void Not_found_예외를_던진다() throws Exception {
            mockMvc.perform(delete("/users/1000"))
                    .andExpect(status().isNotFound());
        }
    }
}
```
![image](https://user-images.githubusercontent.com/45390172/147681594-17d99994-bf3c-42a3-9aa3-ffdf73091705.png)

hannut91 replied:
`@BeforeEach`를 사용할 경우 `given`이 부분이 여러번 실행되는데, 처음에 예외를 던지도록 한 상태에서, 다시 `given`을 하려고 할 때 이미 예외를 던지도록 모킹이 되어있어서 예외를 던져서 문제가 생기는 것 같습니다. 이렇게 사용하셔야 합니다.

```java
doThrow(new UserNotFoundException(1000L))
                    .when(userService).delete(1000L);
```

See also
  - https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#do_family_methods_stubs

Kwakcena replied:
아 그렇군요! 🤭
```java
given(userService.delete(1000L)).willThrow(new UserNotFoundException(1000L));
```
위와 같이 예외를 던지도록 모킹 한 상태에서 다음 테스트에 given이 다시 실행될 때 괄호 안의 `userService.delete(1000L)` 가 실행되므로 예외를 던진 것이군요.

`doThrow`를 보니 void 리턴 타입에 대한 메소드에 대해서 예외를 발생시킬 때 사용하라고 하네요
> Use doThrow() when you want to stub a void method with an exception:

userService의 deleteUser는 void 타입이니 doThrow를 사용하여 잘 해결 되었습니다. 감사합니다.👍

Kwakcena committed:
UserService 를 구현하고 테스트 하라

UserService를 구현하고 이를 테스트 하였습니다

Kwakcena committed:
불필요한 코드 제거

불필요한 코드를 제거합니다

Kwakcena committed:
lombok의 builder를 사용하라

lombok의 getter와 builder를 사용합니다.

Kwakcena committed:
UserData dto를 사용하라

외부와 연결되는 dto인 UserData를 만들었습니다. 순전히 User에 대한 데이터만 갖습니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,152 @@
+        @Test
+        @DisplayName("새로운 회원을 생성하여 리턴한다")
+        void create() {
+            this.createdUser = userService.create(source);
```
createdUser는 멤버 변수로 사용하지 않아도 될 것 같네요

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,152 @@
+
+        @BeforeEach
+        void setUpCreation() {
+            given(userRepository.save(any(User.class))).will(invocation -> {
+                User source = invocation.getArgument(0);
+                return User.builder()
+                        .name(source.getName())
+                        .email(source.getEmail())
+                        .password(source.getPassword())
+                        .build();
+            });
```
```java
private User user;

@BeforeEach 
void setUp() {
    user = UserData.builder()
            .id(1L)
            .name("홍길동")
            .email("test@test.com")
            .password("asdqwe1234")
            .build();
}

@BeforeEach
void before() {
    given(userRepository.save(any(User.class))).willReturn(user);
}

@Test
void create() {
    User createdUser = userService.create(source);

    verify(userRepository).save(any(User.class));

    assertThat(createdUser.getName()).isEqualTo(user.getName());
    assertThat(createdUser.getEmail()).isEqualTo(user.getEmail());
    assertThat(createdUser.getPassword()).isEqualTo(user.getPassword());
}
```

믿고 사용할 수 있는 테스트용 데이터를 만들어놓고 테스트에 사용하면 좋을 것 같아요

Kwakcena replied:
다음의 커밋에 리뷰를 반영 해 보았습니다.
- https://github.com/CodeSoom/spring-week5-assignment-1/pull/49/commits/feb6685601f2a56c8eab4b04e74e2e54dabf8829

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,152 @@
+            }
+        }
+    }
+}
```
마지막 라인에 빈 줄 하나를 추가해 주세요

hannut91 reviewed (commented):
지금은 데이터 자체가 복잡하지 않아서 괜찮아 보일 수 있지만, 테스트용 데이터가 복잡해질수록 테스트도 급격하게 복잡해집니다.
따라서 테스트용 데이터를 어떻게 관리할지 고민해 보시면 좋을 것 같아요

Kwakcena committed:
DozerMapper를 사용하라

DozerMapper를 적용하고 이를 UserService에서 create 하는 곳에 사용 해 봤습니다

Kwakcena commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -54,7 +57,7 @@ void setUpCreation() {
         @DisplayName("새로운 회원을 생성하여 리턴한다")
         void create() {
-            this.createdUser = userService.create(source);
+            createdUser = userService.create(source);
```
UserService의 create 메소드가 Dozer를 사용하도록 고치고 난 뒤 이 부분의 테스트가 깨졌습니다.
에러는 다음과 같이 나타났습니다.
![image](https://user-images.githubusercontent.com/45390172/147766495-ae0c2c37-7573-4857-b868-f527424f6f06.png)

`java.lang.NoSuchMethodException: com.codesoom.assignment.domain.User.<init>()` 에 대해서 검색 해 보니, 어떠한 parameters도 받지 않는 기본 생성자를 누군가가 호출할 때 발생한다고 하더군요.

User 에는 기본 생성자가 없는데 이를 호출하려고 하니 문제가 되었던 것 같습니다.

Kwakcena committed:
명시적인 User 도메인의 생성자를 만들어라

AllArgsConstructor 어노테이션은 사용하지 않는게 좋다는 종립님의 리뷰를 본 적이 있는데, 이를 사용하지 않으면 User 도메인을 DoserMapper로 사용하지 못하는 이슈가 있었습니다.

그래서 명시적으로 생성자를 만들어 생성자에 Builder를 붙였습니다.
AllArgsConstructor를 왜 사용하면 안좋은지는 알아봐야겠습니다.

Kwakcena committed:
테스트 코드에 리뷰를 반영하라

일부 테스트에 사용되는 데이터는 신뢰할 수 있는 데이터를 만들어 놓고
사용하도록 합니다.

Kwakcena committed:
void return type의 예외는 doThrow를 사용한다

void return type을 갖는 메소드가 예외를 던지는 상황에 대해
테스트를 할때는 doThrow를 사용할 수 있다

Kwakcena committed:
deprecated된 APPLICATION_JSON_UTF8을 제거한다

mockMvc 멤버 변수에 대한 정의를 할 때
CharacterEncodingFilter를 주어 해결할 수 있다.

see also: https://github.com/Kyuwon53/Kyuwon53.github.io/blob/main/Spring/%5BSpring%5DMediaType.APPLICATION_JSON_UTF8%EB%A7%90%EA%B3%A0%20%EC%9D%B8%EC%BD%94%EB%94%A9.md

Kwakcena committed:
CrudRepository를 상속하라

UserRepository가 CrudRepository를 상속하지 않아서 서버가 제대로 실행되지 않아
CI가 깨집니다. 이를 수정하였습니다.

Kwakcena committed:
일부 정보만 변경이 가능하게 하라

사용자 정보를 update 할 때는 일부 정보만 변경이 가능할 수
있기 떄문에 Valid를 쓰면 안된다. 전체 정보에 대한 요청이
아니면 에러를 반환하기 떄문이다.

Kwakcena committed:
RroductRepository도 Crud를 상속하라

ProductRepository도 CrudRepository를 상속하도록 합니다

Kwakcena committed:
CI가 통과하도록 수정하라

user.spec.js 에서 update user의 데이터를 수정합니다

Kwakcena committed:
UserController의 javadoc을 작성하라

UserController에 javadoc을 작성하였습니다.

Kwakcena committed:
UserService에 javadoc을 작성하라

UserService에 javadoc을 작성하였습니다.

Kwakcena committed:
userService의 delete메소드 명을 수정하라

userService의 delete 메소드 이름을 deleteUserById로 수정하였습니다.

Kwakcena committed:
사용하지 않는 코드 제거

사용하지 않는 코드를 제거합니다

hannut91 commented on app/src/test/java/com/codesoom/assignment/application/UserServiceTest.java:
```diff
@@ -0,0 +1,173 @@
+        @BeforeEach
+        void setUpCreation() {
+            given(userRepository.save(any(User.class))).willReturn(user);
+        }
```
이 둘이 실행되는게 순서를 보장하지 않을거에요. 다음과 같이 쓰던지 혹은 다음과 같이 써야 할 것 같아요.
```java
        @BeforeEach
        void setUp() {
            source = UserData.builder()
                    .name("홍길동")
                    .email("test@test.com")
                    .password("asdqwe1234")
                    .build();

            user = mapper.map(source, User.class);

            setUpCreation();
        }

        void setUpCreation() {
            given(userRepository.save(any(User.class))).willReturn(user);
        }
```

hannut91 reviewed (commented):
과제풀이를 통해서 꼭 복습해 주세요

Kwakcena committed:
UserRegistration Dto를 만들고 활용한다

회원가입에 쓰이는 정보를 담당하는 UserRegistration DTO를 만들고 이를 활용합니다

Kwakcena committed:
이미 존재하는 이메일에 대해 에러를 던진다

이미 존재하는 이메일로 가입하려 할 경우 에러를 던집니다

hannut91 commented:
이번 주 과제를 통해서 이전과 비교하여 무엇을 할 수 있게 되셨나요?

Kwakcena committed:
UserResultData dto를 만들어 사용한다

응답으로 돌려 줄 사용자의 dto를 UserResultData로 정의하여 사용합니다

Kwakcena committed:
이미 존재하는 이메일에 대한 예외 처리 추가

빼먹고 커밋하지 않았던 예외를 추가합니다.

Kwakcena committed:
UserModificationData dto를 사용하라

updateUser 가 실행될 때 requestBody로 들어오는 데이터에 대해서 UserModificationData를 통해 들어도록 합니다.

Kwakcena committed:
User test를 추가하고 changeWith를 사용한다

UserTest를 추가하고 changeWith라는 메소드를 사용하도록 수정합니다.

Kwakcena committed:
destroy기능 추가

Kwakcena committed:
findUser 메소드로 분리하라

id를 통해서 user를 찾는 메소드 내부에 user.isDeleted를 확인하여 삭제된 회원인 경우에도 에러를 던지도록 합니다

Kwakcena committed:
FindByIdAndDeletedIsFalse 사용

UserRepository에 findByIdAndDeletedIsFalse를 사용하여 삭제되지 않은 회원중에서 찾도록 한다

Kwakcena committed:
getUserResultData로 추출하기

getUserResultData 함수를 만들어서 중복되는 부분을 추출합니다

Kwakcena committed:
정정 사항 반영

existsByEmail의 반환 결과가 에러가 아닌 boolean 데이터이므로 mocking을 수정합니다
